lucene/core/src/java/org/apache/lucene/store/EncryptingDirectory.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.store;
+
+import java.io.IOException;
+
+import org.apache.lucene.index.SegmentInfo;
+
+public class EncryptingDirectory extends FilterDirectory {
+
+  private final KeySupplier keySupplier;
+  private final SegmentKeySupplier segmentKeySupplier;
+  private final SegmentInfo segmentInfo;
+
+  public EncryptingDirectory(Directory directory, KeySupplier keySupplier) {
```

Comment:
you can create two objects: one with the keySupplier and the directory, the second one with the segment key supplier and segment info and instead of using the constructors use an static factory method

Reply:
Right, I'll split it into FileEncryptingDirectory and SegmentEncryptingDirectory.